File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Item.java
package com.hello.uxml.tools.framework;

import java.util.EnumSet;

/**
 * Implements a ContentContainer for items in ListBox/Tree controls.
 *
 * @author ferhat
 */
public class Item extends ContentContainer{

  public static PropertyDefinition selectedPropDef =
      ListBase.selectedPropDef;

  public static PropertyDefinition isFirstPropDef =
      ItemsContainer.isFirstPropDef;

  public static PropertyDefinition isLastPropDef =
      ItemsContainer.isLastPropDef;

  /** IsPressed property definition */
  public static PropertyDefinition isPressedPropDef = PropertySystem.register("IsPressed",
      Boolean.class, Item.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  /**
   * Returns selection state of element.
   */
  public boolean getSelected() {
    return ListBase.getChildSelected(this);
  }

  /**
   * Returns true if item is first item in an ItemsContainer.
   */
  public boolean getIsFirst() {
    return ItemsContainer.getChildIsFirst(this);
  }

  /**
   * Returns true if item is last item in an ItemsContainer.
   */
  public boolean getIsLast() {
    return ItemsContainer.getChildIsLast(this);
  }

  /**
   * Returns whether mouse button over an item is in pressed state.
   */
  public boolean getIsPressed() {
    return (Boolean) getProperty(isPressedPropDef);
  }

  /**
   * Sets selection state of element.
   */
  public void setSelected(boolean selected) {
    ListBase.setChildSelected(this, selected);
  }

  /**
   * Returns parent Item of element.
   */
  public static Item getParentItem(UIElement element) {
    while (element != null) {
      if (element instanceof Item) {
        return (Item) element;
      }
      element = element.getParent();
    }
    return null;
  }
}

File: framework/lib/item.dart
part of uxml;

/**
 * Implements a ContentContainer for listbox/tree items.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class Item extends ContentContainer {
  // This class could also be called SelectableRepeatableContent
  // since it manages the selection state of an item and has chrome to show
  // selection.
  static ElementDef itemElementDef;
  /** IsPressed property definition */
  static PropertyDefinition isPressedProperty;

  /**
   * Constructor.
   */
  Item([Object content = null]) : super() {
    if (content != null) {
      setProperty(ContentContainer.contentProperty, content);
    }
  }

  /** @see UIElement.initSurface. */
  void initSurface(UISurface parentSurface, [int index = -1]) {
    super.initSurface(parentSurface, index);
    if (_hostSurface != null) {
      _hostSurface.hitTestMode = UISurfaceImpl.HITTEST_BOUNDS;
    }
  }

  /**
   * Sets or returns if item is selected.
   */
  bool get selected => ListBase.getChildSelected(this);

  set selected(bool value) {
    ListBase.setChildSelected(this, value);
  }

  /**
   * Returns if item is first in ItemsContainer.
   */
  bool get isFirst => ItemsContainer.getChildIsFirst(this);

  /**
   * Returns if item is last in ItemsContainer.
   */
  bool get isLast => ItemsContainer.getChildIsLast(this);

  /** Overrides UIElement.onMouseDown to update IsPressed property. */
  void onMouseDown(MouseEventArgs mouseArgs) {
    captureMouse();
    setProperty(isPressedProperty, true);
    mouseArgs.handled = true;
  }

  /**
   * Returns true if button is currently in pressed state.
   */
  bool get isPressed {
    return getProperty(isPressedProperty);
  }

  /** Overrides UIElement.onMouseUp to set item selected property. */
  void onMouseUp(MouseEventArgs mouseArgs) {
    /** Overrides UIElement.onMouseDown. */
    if (isPressed) {
      releaseMouse();
      setProperty(isPressedProperty, false);
      if (isMouseOver) {
        selected = true;
      }
    }
  }

  /**
   * Returns parent Item of element.
   */
  static Item getParentItem(UIElement element) {
    while (element != null) {
      if (element is Item) {
        return element;
      }
      element = element.parent;
    }
    return null;
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => itemElementDef;

  /** Registers component. */
  static void registerItem() {
    isPressedProperty = ElementRegistry.registerProperty("isPressed",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    itemElementDef = ElementRegistry.register("Item",
        ContentContainer.contentcontainerElementDef, [isPressedProperty], null);
  }
}
